src/groups/bmq/bmqc/bmqc_multiqueuethreadpool.h

```diff
@@ -1158,23 +1172,46 @@ inline void MultiQueueThreadPool<TYPE>::stop()
     // destructor, then we should be able to just disable the queues in the
     // loop below and let the Queues handle failing the enqueue if we're
     // stopped/stopping.
+
+    // 1. Request queues to stop
     for (size_t i = 0; i < d_queues.size(); ++i) {
         QueueInfo& info = d_queues[i];
 
-        QueueItem item = {false  // isMonitorEvent
-                          ,
-                          0};  // pointer to event
-        info.d_queue_p->pushBack(item);
+        // According to `d_processQueueRefCount` usage contract,
+        // we have to apply the following updates here:
+        // (1) Since we enqueue the next monitor event to the queue:
+        //     `info.d_processQueueRefCount.addRelaxed(1);`
+        // (2) Since we are getting rid of the initial reference:
+        //     `info.d_processQueueRefCount.subtractRelaxed(1);`
+        //
+        // These two updates balance each other (+1 -1 = 0), so we can keep
+        // the current value of `info.d_processQueueRefCount` unchanged.
+
+        info.d_queue_p->pushBack(NULL /* monitor event */);
+        // It is possible that something is enqueued to the queue between the
+        // last monitor event and `disablePushBack()` call, this is expected.
         info.d_queue_p->disablePushBack();
+    }
 
-        // TODO: Change 'd_processedZero' to semaphore instead to avoid busy
-        //       waiting?
-        while (!info.d_processedZero) {
-            bslmt::ThreadUtil::yield();
+    // 2. Verify that queues have stopped
+    const bsls::TimeInterval timeout =
+        bsls::SystemTime::nowMonotonicClock().addSeconds(
+            k_MAX_WAIT_SECONDS_AT_SHUTDOWN);
+    for (size_t i = 0; i < d_queues.size(); ++i) {
+        QueueInfo& info = d_queues[i];
+
+        const int rc = info.d_finished.timedWait(timeout);
+        if (0 != rc) {
+            BALL_LOG_ERROR << "#MQTP_STOP_FAILURE MQTP failed to stop in "
+                           << k_MAX_WAIT_SECONDS_AT_SHUTDOWN
+                           << " seconds while shutting down the queue (" << i
+                           << ", " << info.d_name << ", " << info.d_queue_p
+                           << "), rc:  " << rc << ".";
```

Comment:
I also think that it's better to just crash in this scenario, not trying to do anything else gracefully. It might be dangerous to continue.
We either have a controlled crash here or some undefined operations and a crash in a random place later.